File: online-assessment/arcbest/pid.py
class PIDController:
    def __init__(self, kp, ki, kd, setpoint=0.0):
        """
        Initialize the PID controller with specified gains and setpoint.
        :param kp: Proportional gain
        :param ki: Integral gain
        :param kd: Derivative gain
        :param setpoint: Desired target value
        """
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.setpoint = setpoint

        self.prev_error = 0
        self.integral = 0

    def update(self, current_value, dt):
        """
        Update the PID controller and calculate the control output.
        :param current_value: The current process value
        :param dt: Time step (in seconds)
        :return: Control output
        """
        # Calculate error
        error = self.setpoint - current_value

        # Proportional term
        p = self.kp * error

        # Integral term
        self.integral += error * dt
        i = self.ki * self.integral

        # Derivative term
        derivative = (error - self.prev_error) / dt if dt > 0 else 0
        d = self.kd * derivative

        # Compute total output
        output = p + i + d

        # Update previous error
        self.prev_error = error

        return output


# Example usage
if __name__ == "__main__":
    # Define PID gains
    kp = 1.0
    ki = 0.5
    kd = 0.1

    # Create a PID controller
    pid = PIDController(kp, ki, kd, setpoint=10.0)

    # Simulate a process
    current_value = 0.0
    dt = 0.1  # Time step in seconds
    for t in range(100):  # Simulate for 10 seconds (100 steps)
        output = pid.update(current_value, dt)
        current_value += output * dt  # Simulate process response
        print(f"Time: {t * dt:.1f}s, Output: {output:.2f}, Current Value: {current_value:.2f}")

File: online-assessment/arcbest/pid.cpp
#include <iostream>
#include <chrono>
#include <thread>

class PIDController {
private:
    double kp; // Proportional gain
    double ki; // Integral gain
    double kd; // Derivative gain

    double prevError;      // Error from the previous step
    double integral;       // Accumulated integral
    double prevTime;       // Previous time in seconds

public:
    // Constructor
    PIDController(double kp, double ki, double kd)
        : kp(kp), ki(ki), kd(kd), prevError(0), integral(0), prevTime(0) {}

    // Compute PID output
    double compute(double setpoint, double current, double dt) {
        // Calculate error
        double error = setpoint - current;

        // Calculate integral
        integral += error * dt;

        // Calculate derivative
        double derivative = (error - prevError) / dt;

        // PID output
        double output = (kp * error) + (ki * integral) + (kd * derivative);

        // Update state
        prevError = error;

        return output;
    }
};

int main() {
    // PID coefficients
    double kp = 1.0;
    double ki = 0.5;
    double kd = 0.1;

    PIDController pid(kp, ki, kd);

    // Simulation variables
    double setpoint = 100.0; // Desired value
    double processValue = 0.0; // Current value of the process
    double dt = 0.1; // Time step (in seconds)

    std::cout << "Time\tSetpoint\tProcess Value\tControl Output\n";

    // Simulation loop
    for (int i = 0; i < 100; ++i) {
        double controlOutput = pid.compute(setpoint, processValue, dt);

        // Simulate process dynamics (example: system response)
        processValue += controlOutput * dt;

        // Print the results
        std::cout << i * dt << "\t" << setpoint << "\t\t" << processValue << "\t\t" << controlOutput << "\n";

        // Wait for the next cycle
        std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(dt * 1000)));
    }

    return 0;
}
